### ISSUE TITLE
Layer not working locally due to not supporting zip file

### ISSUE THREAD

**Issue by GeekOnGadgets** [aws/aws-sam-cli#947]
**Description:**

I have a lambda layer which takes care of common code and node_modules, which works perfectly fine when deployed. But for local development its not working. Not sure if, I am doing something wrong or this feature is not supported yet. 

Below is `dir` structure

```
-src/
--test
---index.js
--shared
---nodejs
----package.json
----response.js       
-template.yaml
-.gitignore
```

template.yaml
```
AWSTemplateFormatVersion: '2010-09-09'
Transform: 'AWS::Serverless-2016-10-31'

Globals:
  Function:
    Runtime: nodejs8.10

Parameters:
  StageName:
    Type: String

Resources:
  TestApi:
    Type: AWS::Serverless::Api
    Properties:
      StageName: dev

  #Lambda Layer for sharing code
  TestLambdaLayer:
    Type: AWS::Serverless::LayerVersion
    Properties:
      LayerName: test-layer
      Description: Lambda layer with all node module dependency and shared code between lambda
      ContentUri: src/shared/nodejs.zip
      CompatibleRuntimes:
        - nodejs8.10
      RetentionPolicy: Retain

  Test:
    Type: 'AWS::Serverless::Function'
    Properties:
      Layers:
        - !Ref TestLambdaLayer
      FunctionName: test
      CodeUri: src/test
      Handler: index.test
      Events:
        EndPoint:
          Type: Api
          Properties:
            RestApiId: !Ref TestApi
            Path: /test
            Method: get
```
test api
```
const sharedPath = process.env.LAYER_PATH || '/opt/nodejs/';

const boom = require('boom');

const { successResponse, errorResponse } = require(sharedPath + 'response');

async function test(event) {
  const data = {
    buildNumber: 1.0
  };
  try {
    return successResponse(data);
  } catch (error) {
    return errorResponse(boom.serverUnavailable());
  }
}

module.exports = {
  test
};

```

when I run `sam local start-api` i get `Unable to import module 'index': Error` . When I put node_module and shared code inside `test`  folder path everything works as expected locally. 

How can I make lambda layer work locally without changing the code? Thanks for the help. 

P.S: Also if you have some time I have another issue open on StackOverflow
https://stackoverflow.com/questions/54261625/api-gateway-options-method-throwing-403/54262046#54262046

**Comment by brandond** [comment]
I just ran into this myself. The issue appears to be due to the fact that the layer builder does not actually extract out the layer archives as it does the function code. This means that [create_tar](https://github.com/awslabs/aws-sam-cli/blob/develop/samcli/lib/utils/tar.py#L28) adds the zipfile to the tar as LAYERNAME, and when [_generate_dockerfile](https://github.com/awslabs/aws-sam-cli/blob/develop/samcli/local/docker/lambda_image.py#L224) adds it to the image the still-compressed zipfile is just dropped into /opt/ as-is.

It appears that the layer CodeUri is expected to be a directory, not a zip file - which is unexpected, since the lambda function CodeUri can be either and is extracted out on demand.

**Comment by daveykane** [comment]
+1

**Comment by luissilvaos** [comment]
+1

**Comment by ratajczak** [comment]
I just run in to this when I tried to debug lambda with layers locally.
It's really disappointing that this is already open for 18 months...

The minimum what should be done asap is to add this issue reference to this page:
https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-using-debugging-python.html

And clearly state that debugging with layers doesn't work locally because of this issue.
It would save me few hours of getting to the bottom of this on my own.

**Comment by byrro** [comment]
This is extremely frustrating... The only workaround available is manually copying the Layer code into the Function directory, or is there something better? This approach complete defeats the purpose of using a Layer in the first place. I can't undertand why AWS would publish a Layer service with a half-baked developer tool that doesn't work and things stay that way for ~2 years!

**Comment by jmkanagy** [comment]
I just wasted hours of my life on this that I will never get back.

**Comment by singledigit** [comment]
Layers do work with local development with both the full structure as well as just the manifest file and including a build method in the SAM template. My recommendation is the second method. I talk through how to do this in my Sessions with SAM series. Specifically this episode: https://youtu.be/o72i8_FH320. Please let me know if you need more help.

**Comment by ratajczak** [comment]
@singledigit it doesn't work when you debug locally and use zipped layer.
it's explained here:
https://github.com/aws/aws-sam-cli/issues/947#issuecomment-456567168
When you use zipped file
e.g. ContentUri: some-layer.zip
and you try to debug your function locally sam local creates a container and mounts this file as a folder in /opt/ in the container without decompressing it, just removes the file extension and mounts it like it was a folder not a file.
The simplest way to reproduce that is to use a zipped layer, start debugging and start bash in this container and try to change directory to /opt/some-layer. You won't be able to do this because it will be a file not a directory

**Comment by ehansen31** [comment]
Is there a work around to the zip extraction issue to get `sam local invoke` to run and import a layer properly?

**Comment by s-shibasaki** [comment]
If you want a quick fix,

1. Change the TestLambdaLayer to the following

```yaml
TestLambdaLayer:
  Type: AWS::Serverless::LayerVersion
  Properties:
    LayerName: test-layer
    Description: Lambda layer with all node module dependency and shared code between lambda
    ContentUri: src/shared/ #Modified
    CompatibleRuntimes:
      - nodejs8.10
    RetentionPolicy: Retain
  Metadata: #Added
    BuildMethod: makefile
```

2. Add the following file to src/shared with the name `Makefile`.

```makefile
build-TestLambdaLayer:
    unzip nodejs.zip -d "$(ARTIFACTS_DIR)"
```

Hope it works well.

**Comment by mgrandis** [comment]
Released in [1.24.0](https://github.com/aws/aws-sam-cli/releases/tag/v1.24.0)

**Comment by stwr667** [comment]
> It appears that the layer CodeUri is expected to be a directory, not a zip file - which is unexpected, since the lambda function CodeUri can be either and is extracted out on demand.

Came across this issue today for Java. Except in Java, `CodeUri` **must** be a zip file. I used a directory, and it isn't supported for Java. See working example here: https://github.com/awsdocs/aws-lambda-developer-guide/blob/main/sample-apps/s3-java/template.yml#L43
